src/Featurizers/LagLeadOperatorFeaturizer.h

```diff
@@ -0,0 +1,224 @@
+        *std::min_element(_offsets.cbegin(), _offsets.cend()) < 0 ?  *std::min_element(_offsets.cbegin(), _offsets.cend()) : 0
+    )),
+    // if there are more than one offset provided, we need find the difference between max and min value to determine buffer size
+    // and if there's only one offset, we use the absolute value of it as the size
+    _buffer(
+        _horizon + 
+        (_offsets.size() == 1 ? 
+        static_cast<size_t>(std::abs(_offsets[0])) : 
+        static_cast<size_t>(_max_future_offset-_max_prior_offset))
+    ) {
+        // prepopulate circular buffer with null values to imitate non-existing prior rows
+        for (std::int64_t i = 0; i < (static_cast<std::int64_t>(_horizon) - _max_prior_offset - 1); ++i) {
+            _buffer.push(Traits<nonstd::optional<T>>::CreateNullValue());
+        }
+}
+
+template <typename T>
+LagLeadOperatorTransformer<T>::LagLeadOperatorTransformer(Archive &ar) :
+    LagLeadOperatorTransformer(
+        [&ar](void) {
+            // Version
+            std::uint16_t                   majorVersion(Traits<std::uint16_t>::deserialize(ar));
+            std::uint16_t                   minorVersion(Traits<std::uint16_t>::deserialize(ar));
+
+            if(majorVersion != 1 || minorVersion != 0)
+                throw std::runtime_error("Unsupported archive version");
+
+            // Data
+            std::uint32_t horizon = Traits<std::uint32_t>::deserialize(ar);
+            std::vector<std::int64_t> offsets = Traits<std::vector<std::int64_t>>::deserialize(ar);
+            
+            return LagLeadOperatorTransformer(std::move(horizon), std::move(offsets));
+        }()
+    ) {
+}
+
+template <typename T>
+bool LagLeadOperatorTransformer<T>::operator==(LagLeadOperatorTransformer const &other) const {
+    return this->_horizon == other._horizon
+            && this->_offsets == other._offsets;
+}
+
+template <typename T>
+bool LagLeadOperatorTransformer<T>::operator!=(LagLeadOperatorTransformer const &other) const {
+    return (*this == other) == false;
+}
+
+template <typename T>
+void LagLeadOperatorTransformer<T>::save(Archive &ar) const /*override*/ {
+    // Version
+    Traits<std::uint16_t>::serialize(ar, 1); // Major
+    Traits<std::uint16_t>::serialize(ar, 0); // Minor
+
+    // Data
+    Traits<std::uint32_t>::serialize(ar, _horizon);
+    Traits<std::vector<std::int64_t>>::serialize(ar, _offsets);
+
+    // Note that we aren't serializing working state
+}
+
+template <typename T>
+void LagLeadOperatorTransformer<T>::execute_helper(nonstd::optional<typename BaseType::InputType> const &input, typename BaseType::CallbackFunction const &callback) {
+    _buffer.push(input);
+    // until the circular buffer is full, we don't have enough data to generate output matrix
+    if (_buffer.is_full()) {
+        Microsoft::Featurizer::RowMajMatrix<nonstd::optional<T>> ret(_offsets.size(), _horizon);
+        for (std::uint32_t row = 0; row < _offsets.size(); ++row) {
```

Comment:
>row [](start = 27, length = 3)

Rather than calling this `row`, call it `offset` as that is the language that you have already established with the class' constructor and member variables. That will eliminate the need for the comment on line 209.

Reply:
I think I introduce the concept of `row` because I need `row` to populate the output matrix. If I call it `offset`, wouldn't it be confusing for somebody to read `output_matrix[offset, col] = ...` ?